### PR TITLE
upstream: fix for race condition in tcp conn management

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
fix for race condition in tcp conn management

Additional Description:
I discovered this race in very simple setup:
`2 clients -> [listener->tcp_proxy->cluster->endpoint] -> server`
clients runs in a loop and open short lived tcp sessions. 
Envoy debug build triggers [assertion](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/conn_pool/conn_pool_base.cc#L38). Release build does not crash, but as explained below, it may lead to unexpected behaviour. 

Here is analysis of the problem. As with any race condition it is hard to describe without going to details.
The root cause of the race is that when upstream tcp session terminates, [ConnPoolImplBase::checkForIdleAndNotify](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/conn_pool/conn_pool_base.cc#L451-L459) is run twice.
 
It is run first time when Local/RemoteClose happens in https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/conn_pool/conn_pool_base.cc#L549.
[ConnPoolImplBase::checkForIdleAndNotify](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/conn_pool/conn_pool_base.cc#L451-L459) iterates through registered idle_callbacks_. The callback is a [lambda with captured host and hash_key](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/upstream/cluster_manager_impl.cc#L2118-L2120) which removes the tcp connection pool: [source/common/upstream/cluster_manager_impl.cc](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/upstream/cluster_manager_impl.cc#L2136-L2141).

The second time ConnPoolImplBase::checkForIdleAndNotify is called from [Envoy::Tcp::ActiveTcpClient::~ActiveTcpClient](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/tcp/conn_pool.cc#L51) via [ConnPoolImplBase::checkForIdleAndCloseIdleConnsIfDraining](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/conn_pool/conn_pool_base.cc#L466). It again will go through registered idle callbacks. It will try to remove the tcp connection pool based on hash_key, but it was already removed when the [ConnPoolImplBase::checkForIdleAndNotify](https://github.com/envoyproxy/envoy/blob/release/v1.28/source/common/conn_pool/conn_pool_base.cc#L451-L459) was called the first time, so it is no-op.

The problem happens when a new connection comes between first and second call to ConnPoolImplBase::checkForIdleAndNotify. The new connection will allocate a new tcp connection pool with the same hash_key as the one just deleted in the first call to ConnPoolImplBase::checkForIdleAndNotify. Then, the second call to ConnPoolImplBase::checkForIdleAndNotify invokes registered callbacks the second time. The callback basically tries to delete a connection pool indexed by hash_key. It finds the newly allocated session which was created between calls to  ConnPoolImplBase::checkForIdleAndNotify.

 There are few ways to make sure that callbacks are run only once when upstream tcp session terminates. A boolean flag is one possibility or clearing the list of callbacks after they have been called.

@alyssawilk @ggreenway 

Risk Level: Low
Testing: Manual. Before the fix assertion triggered within 5 minutes. After the fix crash does not happen.
Docs Changes: No
Release Notes: No
Platform Specific Features: No
Fixes: #22583

### CODE REVIEW

**Review by ggreenway — CHANGES_REQUESTED** [review]
Thank you for the detailed description of what's happening!

I think the fix you have here makes sense and is straightforward.

Is it possible to create a test to validate that the idle callback cannot be called multiple times?

Also, can you add a comment with your analysis of this bug next to the fix so a future reader understands the issue? I don't want this knowledge to be lost.

/wait

**Comment by ggreenway** [comment]
@alyssawilk can you also take a look at this fix and double-check that it is the correct fix?

**Comment by cpakulski** [comment]
@ggreenway thanks for reviewing. I will try add a unit test to make sure that callbacks are called exactly once regardless of path taken (Local/RemoteClose) or destructor.

**Review by alyssawilk — DISMISSED** [review]
LGTM though yeah it'd be good to have a unit test especially as I'm not sure if there's a non-bug situation where it should get called twice so we may just be masking a larger issue.

**Comment by kyessenov** [comment]
This is causing #22583 and https://github.com/istio/proxy/pull/5151.

**Comment by cpakulski** [comment]
I will finish this PR by Monday and we can do final review.

**Comment by cpakulski** [comment]
Updated PR description with "Fixes ...". Thanks for linking those issues @kyessenov.

File: source/common/conn_pool/conn_pool_base.cc
Comment (kyessenov) [inline source/common/conn_pool/conn_pool_base.cc:459]: My fix was to put 
```c++
if (deferred_deleting_) {
  return;
}
```
since it was protecting against future accidents. Either way SGTM. Can I suggest to change the debug log to indicate whether it's firing or not? It's useful for debugging.

Comment (cpakulski) [reply source/common/conn_pool/conn_pool_base.cc:459]: Yeah, makes sense. I will change the log.

**Comment by cpakulski** [comment]
I have added new unit test to verify that callbacks are called exactly once regardless whether upstream connection has been fully established (Connected, Local/Remote Reset events) or partially created (ActiveTcp client constructor and destructor have been called only).

I think this PR is finished and ready for the final review.

File: test/common/conn_pool/conn_pool_base_test.cc
Comment (alyssawilk) [inline test/common/conn_pool/conn_pool_base_test.cc:483]: Just to check this failed befre?  not sure if with nice mocks calling twice would pass.  If so, LGTM

**Review by alyssawilk — APPROVED** [review]
hm, it's a clear improvement and I guess as long as it's tested

**Comment by alyssawilk** [comment]
@ggreenway has my LGTM but auto-merge blocked by your requested changes.  resolve when you have a few?

**Comment by zirain** [comment]
@alyssawilk need backport?

**Comment by alyssawilk** [comment]
I'd say if this fixes istio's issue then yes.  They said they'd confirm this/next week

**Pull request merged**
